### PR TITLE
FIX_1061

### PR DESCRIPTION
**Pull request by franchuterivera**
Addresses #1061 

To properly use crossvalidate method, we need to comply with autosklearn classification requirements as stated in https://scikit-learn.org/dev/glossary.html#term-classes

This PR enhances the default estimator with the missing components. Also, it addresses a problem with joblib in which we do not comply with set/get params requirement that the signature should be stored in self.

### CODE REVIEW

File: test/test_automl/test_estimators.py [hunk 1]
```diff
@@ -275,6 +289,27 @@ def test_cv_results(tmp_dir, output_dir):
                        cv_results.items() if key.startswith('param_')]
     assert all(cv_result_items), cv_results.items()
 
+    # Compare the state of the model parameters with the original parameters
+    new_params = clone(cls).get_params()
```
Comment (mfeurer) [inline test/test_automl/test_estimators.py:293]: Is there a specific reason you added this to the unit test for cv_results?

File: autosklearn/data/target_validator.py [hunk 2]
```diff
@@ -386,3 +386,10 @@ def _check_data(
                                  self.type_of_target,
                                  supported_output_types
                              ))
+
+    @property
+    def classes(self) -> typing.List:
```
Comment (mfeurer) [inline autosklearn/data/target_validator.py:391]: Shouldn't this be `classes_`?

File: test/test_data/test_target_validator.py [hunk 3]
```diff
@@ -461,6 +463,8 @@ def test_targetvalidator_inversetransform():
     y_decoded = validator.inverse_transform(y)
     np.testing.assert_array_almost_equal(y, y_decoded)
 
+    assert validator.classes == []
```
Comment (mfeurer) [inline test/test_data/test_target_validator.py:466]: Shouldn't this be `[[0, 1], [0, 1], [0, 1], [0, 1]]`?

File: test/test_automl/test_estimators.py (hunk 1 above)
Comment (franchuterivera) [reply test/test_automl/test_estimators.py:293]: I really just wanted to test an implementation fo autosklearnclassifier, nothing more. This can be done anywhere else and the reason for that is that clone of the estimator makes sure that the set/get parameters work. Do you think it should be somewhere else?

File: autosklearn/data/target_validator.py (hunk 2 above)
Comment (franchuterivera) [reply autosklearn/data/target_validator.py:391]: This line is the one that needs to be classes https://github.com/automl/auto-sklearn/blob/738f66ccf8f672d56b0c43ca62e9c4d707d2bd48/autosklearn/estimators.py#L604

I can make this function classes_ if you think it makes sense to have the same attribute everywhere.

File: test/test_data/test_target_validator.py (hunk 3 above)
Comment (franchuterivera) [reply test/test_data/test_target_validator.py:466]: For multilabel data we do not perform encoding, because sadly we need a label binarized I think. Else it would be converted (the whole [0, 1, 0] to a single number).

To go in line with the fact that there is no encoding I leave this as empty. We can however ask if the type of target is multilabel indicator, make it so that the classes are the unique elements of the target array. Does this make sense?

File: test/test_automl/test_estimators.py (hunk 1 above)
Comment (mfeurer) [reply test/test_automl/test_estimators.py:293]: No, it's totally fine to have it here.

File: autosklearn/data/target_validator.py (hunk 2 above)
Comment (mfeurer) [reply autosklearn/data/target_validator.py:391]: Well, the idea of `classes_` is to store the classes after fit, adhering to scikit-learn's best practices by having a trailing `_`. As this function does exactly this, yes, I think so.

File: test/test_data/test_target_validator.py (hunk 3 above)
Comment (mfeurer) [reply test/test_data/test_target_validator.py:466]: Hm, I think it's fine this way.

File: test/test_data/test_target_validator.py [hunk 4]
```diff
@@ -450,7 +450,7 @@ def test_targetvalidator_inversetransform():
     y_decoded = validator.inverse_transform(y)
     assert ['a', 'a', 'b', 'c', 'a'] == y_decoded.tolist()
 
-    assert validator.classes.tolist() == ['a', 'b', 'c']
+    assert validator.classes_.tolist() == ['a', 'b', 'c']
```
Comment (mfeurer) [inline test/test_data/test_target_validator.py:453]: Sorry, one more question. According to the type annotation, this is already a list. I guess that the either the type annotation or this check is wrong.

Comment (franchuterivera) [reply test/test_data/test_target_validator.py:453]: Yes it should be np.ndarray to comply with other scikit learn estimators. Sadly mypy ignores numpy (well, treated as ANY) 

There is a typing extension module [here](https://pypi.org/project/typing-extensions/) in case you want to enhance this. For the time being I have moved away from list.

Comment (mfeurer) [reply test/test_data/test_target_validator.py:453]: We're getting proper annotations in the upcoming releases: https://numpy.org/doc/1.20/release/1.20.0-notes.html

Therefore, we can just keep an eye on this and move to these new types once they are stable and scikit-learn uses them, too.

**Review by mfeurer — APPROVED** [review]
